lib/core/scheduler.cpp

```diff
@@ -0,0 +1,39 @@
+#include "scheduler.hpp"
+#include "wall_clock.hpp"
+
+namespace hyped::core {
+
+Scheduler::Scheduler(core::ILogger &logger, core::ITimeSource &time)
+    : logger_(logger),
+      time_(time),
+      task_queue_()
+{
+}
+
+core::Result Scheduler::run()
+{
+  if (task_queue_.empty()) { return core::Result::kSuccess; }
+  const auto current_time = std::chrono::time_point_cast<std::chrono::nanoseconds>(time_.now());
+  auto next_task          = task_queue_.top();
+  core::Result result     = core::Result::kSuccess;
+  for (std::size_t i = 0; i < task_queue_.size(); i++) {
+    if (current_time < next_task.sheduled_time) { return result; }
+    core::Result next_result = next_task.handler();
+    if (next_result == core::Result::kFailure) { result = core::Result::kFailure; }
+    task_queue_.pop();
+    if (task_queue_.empty()) { break; }
+    next_task = task_queue_.top();
+  }
+  return result;
+}
+
+void Scheduler::addTask(const core::Duration delay, std::function<core::Result(void)> handler)
+{
+  core::TimePoint execution_timepoint
+    = std::chrono::time_point_cast<std::chrono::nanoseconds>(time_.now())
+      + std::chrono::nanoseconds(delay);
```

Comment:
this seems clunky. maybe you want to add this sort of functionality inside of `core/time.hpp`?

maybe by adding a `+` operator between `TimePoint` and `Duration`?

Reply:
How does this work? Seeing as both are defined with `using`. I might be being dumb, but I can't see a way to implement e.g.
```c++
Timepoint TimePoint::operator+(Duration const &rhs) const
{
  return TimePoint(this->time_since_epoch() + rhs);
}
```
as it has to be a member of a class

Reply:
You might want to write classes that extend/inherit from the underlying types instead of just aliasing with the `using` keyword. Not sure what what the greater implications here would be, but I suspect that we only need a very small subset of the full functionality so it should be fine to write a wrapper.

Reply:
While this is a bit clunky, I feel like that might be over the top just to avoid this one instance of verbosity